File: stacks/node/base/code/data-validation.tpl.js
/**
 * Template: data-validation.tpl.js
 * Purpose: data-validation template
 * Stack: node
 * Tier: base
 */

# Universal Template System - Node Stack
# Generated: 2025-12-10
# Purpose: Data validation utilities
# Tier: base
# Stack: node
# Category: utilities

#!/usr/bin/env node
/**
 * Node.js Data Validation Template
 * Purpose: Reusable data validation utilities for Node.js projects
 * Usage: Import and adapt for consistent data validation across the application
 */

const { EventEmitter } = require('events');

/**
 * Validation types enumeration
 */
const ValidationType = {
    REQUIRED: 'required',
    STRING: 'string',
    INTEGER: 'integer',
    FLOAT: 'float',
    BOOLEAN: 'boolean',
    EMAIL: 'email',
    PHONE: 'phone',
    URL: 'url',
    DATE: 'date',
    DATETIME: 'datetime',
    JSON: 'json',
    REGEX: 'regex',
    MIN_LENGTH: 'min_length',
    MAX_LENGTH: 'max_length',
    MIN_VALUE: 'min_value',
    MAX_VALUE: 'max_value',
    IN_CHOICES: 'in_choices',
    CUSTOM: 'custom'
};

/**
 * Validation rule class
 */
class ValidationRule {
    constructor(type, options = {}) {
        this.type = type;
        this.params = options.params || {};
        this.message = options.message || null;
        this.required = options.required || false;
    }
}

/**
 * Validation result class
 */
class ValidationResult {
    constructor(field, value) {
        this.field = field;
        this.value = value;
        this.isValid = true;
        this.errors = [];
    }

    /**
     * Add error to validation result
     */
    addError(message) {
        this.errors.push(message);
        this.isValid = false;
    }

    /**
     * Get first error message
     */
    getFirstError() {
        return this.errors.length > 0 ? this.errors[0] : null;
    }

    /**
     * Convert to JSON
     */
    toJSON() {
        return {
            field: this.field,
            value: this.value,
            isValid: this.isValid,
            errors: this.errors
        };
    }
}

/**
 * Data validator class
 */
class DataValidator extends EventEmitter {
    constructor(options = {}) {
        super();
        this.options = {
            strictMode: options.strictMode || false,
            stopOnFirstError: options.stopOnFirstError || false,
            ...options
        };
        this.customValidators = new Map();
        this.globalValidators = new Map();
        this._setupGlobalValidators();
    }

    /**
     * Setup built-in global validators
     */
    _setupGlobalValidators() {
        // Email validator
        this.globalValidators.set('email', (value) => {
            const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
            return emailRegex.test(value);
        });

        // Phone validator
        this.globalValidators.set('phone', (value) => {
            const cleanPhone = value.replace(/[\s\-\(\)]+/g, '');
            const phoneRegex = /^\+?[1-9]\d{9,14}$/;
            return phoneRegex.test(cleanPhone);
        });

        // URL validator
        this.globalValidators.set('url', (value) => {
            try {
                new URL(value);
                return true;
            } catch {
                return false;
            }
        });

        // Date validator
        this.globalValidators.set('date', (value) => {
            if (value instanceof Date) return true;
            const date = new Date(value);
            return !isNaN(date.getTime()) && value.match(/^\d{4}-\d{2}-\d{2}$/);
        });

        // DateTime validator
        this.globalValidators.set('datetime', (value) => {
            if (value instanceof Date) return true;
            const date = new Date(value);
            return !isNaN(date.getTime());
        });

        // JSON validator
        this.globalValidators.set('json', (value) => {
            if (typeof value === 'object') return true;
            try {
                JSON.parse(value);
                return true;
            } catch {
                return false;
            }
        });
    }

    /**
     * Add custom validator
     */
    addCustomValidator(name, validatorFunc, errorMessage = null) {
        this.customValidators.set(name, {
            func: validatorFunc,
            message: errorMessage || `Custom validation '${name}' failed`
        });
    }

    /**
     * Validate a single field
     */
    validateField(fieldName, value, rules) {
        const result = new ValidationResult(fieldName, value);

        for (const rule of rules) {
            const error = this._validateRule(fieldName, value, rule);
            if (error) {
                result.addError(error);
                this.emit('fieldError', { field: fieldName, value, rule, error });
                
                if (this.options.stopOnFirstError) {
                    break;
                }
            }
        }

        this.emit('fieldValidated', result);
        return result;
    }

    /**
     * Validate entire data object
     */
    validate(data, schema) {
        const results = {};
        let hasErrors = false;

        for (const [fieldName, rules] of Object.entries(schema)) {
            const value = data[fieldName];
            results[fieldName] = this.validateField(fieldName, value, rules);
            
            if (!results[fieldName].isValid) {
                hasErrors = true;
            }
        }

        const validationResult = {
            isValid: !hasErrors,
            results,
            errors: this._collectErrors(results)
        };

        this.emit('validationCompleted', validationResult);
        return validationResult;
    }

    /**
     * Validate a single rule
     */
    _validateRule(fieldName, value, rule) {
        const params = rule.params;

        try {
            switch (rule.type) {
                case ValidationType.REQUIRED:
                    if (value === null || value === undefined || value === '') {
                        return rule.message || `Field '${fieldName}' is required`;
                    }
                    break;

                case ValidationType.STRING:
                    if (value !== null && value !== undefined && typeof value !== 'string') {
                        return rule.message || `Field '${fieldName}' must be a string`;
                    }
                    break;

                case ValidationType.INTEGER:
                    if (value !== null && value !== undefined) {
                        if (typeof value !== 'number' || !Number.isInteger(value)) {
                            return rule.message || `Field '${fieldName}' must be an integer`;
                        }
                    }
                    break;

                case ValidationType.FLOAT:
                    if (value !== null && value !== undefined) {
                        const numValue = Number(value);
                        if (isNaN(numValue)) {
                            return rule.message || `Field '${fieldName}' must be a number`;
                        }
                    }
                    break;

                case ValidationType.BOOLEAN:
                    if (value !== null && value !== undefined && typeof value !== 'boolean') {
                        return rule.message || `Field '${fieldName}' must be a boolean`;
                    }
                    break;

                case ValidationType.EMAIL:
                    if (value && !this.globalValidators.get('email')(value)) {
                        return rule.message || `Field '${fieldName}' must be a valid email`;
                    }
                    break;

                case ValidationType.PHONE:
                    if (value && !this.globalValidators.get('phone')(value)) {
                        return rule.message || `Field '${fieldName}' must be a valid phone number`;
                    }
                    break;

                case ValidationType.URL:
                    if (value && !this.globalValidators.get('url')(value)) {
                        return rule.message || `Field '${fieldName}' must be a valid URL`;
                    }
                    break;

                case ValidationType.DATE:
                    if (value && !this.globalValidators.get('date')(value)) {
                        return rule.message || `Field '${fieldName}' must be a valid date (YYYY-MM-DD)`;
                    }
                    break;

                case ValidationType.DATETIME:
                    if (value && !this.globalValidators.get('datetime')(value)) {
                        return rule.message || `Field '${fieldName}' must be a valid datetime`;
                    }
                    break;

                case ValidationType.JSON:
                    if (value && !this.globalValidators.get('json')(value)) {
                        return rule.message || `Field '${fieldName}' must be valid JSON`;
                    }
                    break;

                case ValidationType.REGEX:
                    if (value && params.pattern) {
                        const regex = new RegExp(params.pattern);
                        if (!regex.test(String(value))) {
                            return rule.message || `Field '${fieldName}' does not match required pattern`;
                        }
                    }
                    break;

                case ValidationType.MIN_LENGTH:
                    if (value && String(value).length < params.length) {
                        return rule.message || `Field '${fieldName}' must be at least ${params.length} characters`;
                    }
                    break;

                case ValidationType.MAX_LENGTH:
                    if (value && String(value).length > params.length) {
                        return rule.message || `Field '${fieldName}' must be at most ${params.length} characters`;
                    }
                    break;

                case ValidationType.MIN_VALUE:
                    if (value !== null && value !== undefined) {
                        const numValue = Number(value);
                        if (numValue < params.value) {
                            return rule.message || `Field '${fieldName}' must be at least ${params.value}`;
                        }
                    }
                    break;

                case ValidationType.MAX_VALUE:
                    if (value !== null && value !== undefined) {
                        const numValue = Number(value);
                        if (numValue > params.value) {
                            return rule.message || `Field '${fieldName}' must be at most ${params.value}`;
                        }
                    }
                    break;

                case ValidationType.IN_CHOICES:
                    if (value && params.choices && !params.choices.includes(value)) {
                        return rule.message || `Field '${fieldName}' must be one of: ${params.choices.join(', ')}`;
                    }
                    break;

                case ValidationType.CUSTOM:
                    if (params.validator && this.customValidators.has(params.validator)) {
                        const validator = this.customValidators.get(params.validator);
                        if (!validator.func(value)) {
                            return rule.message || validator.message;
                        }
                    }
                    break;

                default:
                    if (this.options.strictMode) {
                        return rule.message || `Unknown validation type: ${rule.type}`;
                    }
            }
        } catch (error) {
            this.emit('validationError', { fieldName, rule, error });
            return `Validation failed for field '${fieldName}': ${error.message}`;
        }

        return null;
    }

    /**
     * Collect all errors from validation results
     */
    _collectErrors(results) {
        const errors = [];
        
        for (const result of Object.values(results)) {
            if (!result.isValid) {
                errors.push(...result.errors);
            }
        }
        
        return errors;
    }
}

/**
 * Schema builder for fluent validation schema creation
 */
class SchemaBuilder {
    constructor() {
        this.schema = {};
    }

    /**
     * Add field with validation rules
     */
    field(fieldName) {
        return new FieldBuilder(fieldName, this);
    }

    /**
     * Get built schema
     */
    build() {
        return this.schema;
    }

    /**
     * Add field to schema (internal method)
     */
    _addField(fieldName, rules) {
        this.schema[fieldName] = rules;
        return this;
    }
}

/**
 * Field builder for fluent interface
 */
class FieldBuilder {
    constructor(fieldName, schemaBuilder) {
        this.fieldName = fieldName;
        this.schemaBuilder = schemaBuilder;
        this.rules = [];
    }

    required(message = null) {
        this.rules.push(new ValidationRule(ValidationType.REQUIRED, { message }));
        return this;
    }

    string(message = null) {
        this.rules.push(new ValidationRule(ValidationType.STRING, { message }));
        return this;
    }

    integer(message = null) {
        this.rules.push(new ValidationRule(ValidationType.INTEGER, { message }));
        return this;
    }

    float(message = null) {
        this.rules.push(new ValidationRule(ValidationType.FLOAT, { message }));
        return this;
    }

    boolean(message = null) {
        this.rules.push(new ValidationRule(ValidationType.BOOLEAN, { message }));
        return this;
    }

    email(message = null) {
        this.rules.push(new ValidationRule(ValidationType.EMAIL, { message }));
        return this;
    }

    phone(message = null) {
        this.rules.push(new ValidationRule(ValidationType.PHONE, { message }));
        return this;
    }

    url(message = null) {
        this.rules.push(new ValidationRule(ValidationType.URL, { message }));
        return this;
    }

    date(message = null) {
        this.rules.push(new ValidationRule(ValidationType.DATE, { message }));
        return this;
    }

    datetime(message = null) {
        this.rules.push(new ValidationRule(ValidationType.DATETIME, { message }));
        return this;
    }

    json(message = null) {
        this.rules.push(new ValidationRule(ValidationType.JSON, { message }));
        return this;
    }

    regex(pattern, message = null) {
        this.rules.push(new ValidationRule(ValidationType.REGEX, { 
            params: { pattern }, 
            message 
        }));
        return this;
    }

    minLength(length, message = null) {
        this.rules.push(new ValidationRule(ValidationType.MIN_LENGTH, { 
            params: { length }, 
            message 
        }));
        return this;
    }

    maxLength(length, message = null) {
        this.rules.push(new ValidationRule(ValidationType.MAX_LENGTH, { 
            params: { length }, 
            message 
        }));
        return this;
    }

    min(value, message = null) {
        this.rules.push(new ValidationRule(ValidationType.MIN_VALUE, { 
            params: { value }, 
            message 
        }));
        return this;
    }

    max(value, message = null) {
        this.rules.push(new ValidationRule(ValidationType.MAX_VALUE, { 
            params: { value }, 
            message 
        }));
        return this;
    }

    choices(choices, message = null) {
        this.rules.push(new ValidationRule(ValidationType.IN_CHOICES, { 
            params: { choices }, 
            message 
        }));
        return this;
    }

    custom(validatorName, message = null) {
        this.rules.push(new ValidationRule(ValidationType.CUSTOM, { 
            params: { validator: validatorName }, 
            message 
        }));
        return this;
    }

    /**
     * Finish building this field and return to schema builder
     */
    end() {
        this.schemaBuilder._addField(this.fieldName, this.rules);
        return this.schemaBuilder;
    }
}

/**
 * Express middleware for request validation
 */
function createValidationMiddleware(schema, validator = null) {
    const dataValidator = validator || new DataValidator();

    return (req, res, next) => {
        const validationResult = dataValidator.validate(req.body, schema);

        if (!validationResult.isValid) {
            return res.status(400).json({
                error: true,
                message: 'Validation failed',
                errors: validationResult.errors,
                details: validationResult.results
            });
        }

        req.validationResult = validationResult;
        next();
    };
}

/**
 * Predefined validation schemas
 */
function createUserSchema() {
    const builder = new SchemaBuilder();
    
    return builder
        .field('username')
            .required()
            .string()
            .minLength(3)
            .maxLength(50)
            .regex(/^[a-zA-Z0-9_]+$/, 'Username can only contain letters, numbers, and underscores')
            .end()
        .field('email')
            .required()
            .email()
            .end()
        .field('age')
            .integer()
            .min(0)
            .max(150)
            .end()
        .field('status')
            .choices(['active', 'inactive', 'pending'])
            .end()
        .build();
}

function createAPISchema() {
    const builder = new SchemaBuilder();
    
    return builder
        .field('api_key')
            .required()
            .string()
            .minLength(10)
            .end()
        .field('timestamp')
            .datetime()
            .end()
        .field('data')
            .json()
            .end()
        .field('version')
            .regex(/^\d+\.\d+\.\d+$/, 'Version must be in format x.y.z')
            .end()
        .build();
}

/**
 * Utility functions for common validation patterns
 */
function validateEmailList(emails) {
    const validator = new DataValidator();
    const invalidEmails = [];

    for (const email of emails) {
        if (!validator.globalValidators.get('email')(email)) {
            invalidEmails.push(email);
        }
    }

    return invalidEmails;
}

function sanitizeString(value, options = {}) {
    const {
        allowSpaces = true,
        allowSpecial = false,
        maxLength = null
    } = options;

    let pattern;
    if (allowSpaces && allowSpecial) {
        pattern = /[^a-zA-Z0-9\s\-\._@+]/g;
    } else if (allowSpaces) {
        pattern = /[^a-zA-Z0-9\s]/g;
    } else if (allowSpecial) {
        pattern = /[^a-zA-Z0-9\-\._@+]/g;
    } else {
        pattern = /[^a-zA-Z0-9]/g;
    }

    let sanitized = value.replace(pattern, '');
    
    if (maxLength && sanitized.length > maxLength) {
        sanitized = sanitized.substring(0, maxLength);
    }

    return sanitized;
}

function validatePasswordStrength(password) {
    const result = {
        isValid: true,
        score: 0,
        issues: [],
        suggestions: []
    };

    if (password.length < 8) {
        result.isValid = false;
        result.issues.push('Password must be at least 8 characters');
    } else {
        result.score += 1;
    }

    if (!/[a-z]/.test(password)) {
        result.isValid = false;
        result.issues.push('Password must contain lowercase letters');
    } else {
        result.score += 1;
    }

    if (!/[A-Z]/.test(password)) {
        result.isValid = false;
        result.issues.push('Password must contain uppercase letters');
    } else {
        result.score += 1;
    }

    if (!/\d/.test(password)) {
        result.isValid = false;
        result.issues.push('Password must contain numbers');
    } else {
        result.score += 1;
    }

    if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
        result.suggestions.push('Consider adding special characters for stronger security');
    } else {
        result.score += 1;
    }

    // Check for common patterns
    if (/^(.)\1+$/.test(password)) {
        result.issues.push('Password cannot be repeated characters');
        result.isValid = false;
    }

    if (/password|123456|qwerty/i.test(password)) {
        result.issues.push('Password is too common');
        result.isValid = false;
    }

    return result;
}

// Example usage
if (require.main === module) {
    async function main() {
        try {
            // Create validator
            const validator = new DataValidator();

            // Add custom validator
            validator.addCustomValidator('evenNumber', (value) => {
                return typeof value === 'number' && value % 2 === 0;
            }, 'Value must be an even number');

            // Test field validation
            const emailRules = [
                new ValidationRule(ValidationType.REQUIRED),
                new ValidationRule(ValidationType.EMAIL)
            ];

            const result = validator.validateField('email', 'test@example.com', emailRules);
            console.log('Email validation:', result.toJSON());

            // Test schema validation with builder
            const userSchema = createUserSchema();
            const testUser = {
                username: 'john_doe',
                email: 'john@example.com',
                age: 25,
                status: 'active'
            };

            const validation = validator.validate(testUser, userSchema);
            console.log('User validation:', validation);

            // Test password validation
            const passwordResult = validatePasswordStrength('MyPassword123!');
            console.log('Password validation:', passwordResult);

            // Test sanitization
            const sanitized = sanitizeString('Hello World! @#$', { 
                allowSpaces: true, 
                allowSpecial: true 
            });
            console.log('Sanitized string:', sanitized);

            // Test email list validation
            const emails = ['test@example.com', 'invalid-email', 'user@domain.org'];
            const invalidEmails = validateEmailList(emails);
            console.log('Invalid emails:', invalidEmails);

            console.log('Data validation utilities demo completed');

        } catch (error) {
            console.error('Demo error:', error.message);
        }
    }

    main();
}

module.exports = {
    // Classes
    DataValidator,
    ValidationRule,
    ValidationResult,
    SchemaBuilder,
    FieldBuilder,

    // Middleware
    createValidationMiddleware,

    // Utilities
    createUserSchema,
    createAPISchema,
    validateEmailList,
    sanitizeString,
    validatePasswordStrength,

    // Constants
    ValidationType
};

File: stacks/r/base/code/data-validation.tpl.py
# Universal Template System - R Stack
# Generated: 2025-12-10
# Purpose: Data validation utilities
# Tier: base
# Stack: r
# Category: utilities

#!/usr/bin/env r3
"""
R Data Validation Template
Purpose: Reusable data validation utilities for R projects
Usage: Import and adapt for consistent data validation across the application
"""

library(re
library(logging
typing library(Dict, Any, List, Optional, Union, Callable
dataclasses library(dataclass
enum library(Enum
datetime library(datetime, date
library(jsonlite

class ValidationType(Enum):
    """Validation types"""
    REQUIRED = "required"
    STRING = "string"
    INTEGER = "integer"
    FLOAT = "float"
    BOOLEAN = "boolean"
    EMAIL = "email"
    PHONE = "phone"
    URL = "url"
    DATE = "date"
    DATETIME = "datetime"
    JSON = "jsonlite"
    REGEX = "regex"
    MIN_LENGTH = "min_length"
    MAX_LENGTH = "max_length"
    MIN_VALUE = "min_value"
    MAX_VALUE = "max_value"
    IN_CHOICES = "in_choices"
    CUSTOM = "custom"

@dataclass
class ValidationRule:
    """Single validation rule"""
    type: ValidationType
    params: Optional[Dict[str, Any]] = None
    message: Optional[str] = None

@dataclass
class ValidationResult:
    """Validation result"""
    is_valid: bool
    errors: List[str]
    field: str
    value: Any

class ValidationError(Exception):
    """Validation error exception"""
    
    function __init__(self, message: str, field: str = None, value: Any = None, errors: List[str] = None):
        super().__init__(message)
        self.field = field
        self.value = value
        self.errors = errors or []

class DataValidator:
    """Data validation utility"""
    
    function __init__(self, logger: Optional[logging.Logger] = None):
        self.logger = logger or logging.getLogger(__name__)
        self.custom_validators = {}
    
    function add_custom_validator(self, name: str, validator_func: Callable[[Any], bool], error_message: str = None):
        """Add custom validation function"""
        self.custom_validators[name] = {
            'func': validator_func,
            'message': error_message or f"Custom validation '{name}' failed"
        }
    
    function validate_field(self, field_name: str, value: Any, rules: List[ValidationRule]) -> ValidationResult:
        """Validate a single field against rules"""
        errors = []
        
        for rule in rules:
            error = self._validate_rule(field_name, value, rule)
            if error:
                errors.append(error)
        
        return ValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
            field=field_name,
            value=value
        )
    
    function validate_data(self, data: Dict[str, Any], schema: Dict[str, List[ValidationRule]]) -> Dict[str, ValidationResult]:
        """Validate entire data dictionary against schema"""
        results = {}
        
        for field_name, rules in schema.items():
            value = data.get(field_name)
            results[field_name] = self.validate_field(field_name, value, rules)
        
        return results
    
    function _validate_rule(self, field_name: str, value: Any, rule: ValidationRule) -> Optional[str]:
        """Validate a single rule"""
        params = rule.params or {}
        
        try:
            if rule.type == ValidationType.REQUIRED:
                if value is None or value == '':
                    return rule.message or f"Field '{field_name}' is required"
            
            elif rule.type == ValidationType.STRING:
                if value is not None and not isinstance(value, str):
                    return rule.message or f"Field '{field_name}' must be a string"
            
            elif rule.type == ValidationType.INTEGER:
                if value is not None:
                    if not isinstance(value, int) or isinstance(value, bool):
                        return rule.message or f"Field '{field_name}' must be an integer"
            
            elif rule.type == ValidationType.FLOAT:
                if value is not None:
                    try:
                        float(value)
                    except (ValueError, TypeError):
                        return rule.message or f"Field '{field_name}' must be a number"
            
            elif rule.type == ValidationType.BOOLEAN:
                if value is not None and not isinstance(value, bool):
                    return rule.message or f"Field '{field_name}' must be a boolean"
            
            elif rule.type == ValidationType.EMAIL:
                if value and not self._is_valid_email(value):
                    return rule.message or f"Field '{field_name}' must be a valid email"
            
            elif rule.type == ValidationType.PHONE:
                if value and not self._is_valid_phone(value):
                    return rule.message or f"Field '{field_name}' must be a valid phone number"
            
            elif rule.type == ValidationType.URL:
                if value and not self._is_valid_url(value):
                    return rule.message or f"Field '{field_name}' must be a valid URL"
            
            elif rule.type == ValidationType.DATE:
                if value and not self._is_valid_date(value):
                    return rule.message or f"Field '{field_name}' must be a valid date"
            
            elif rule.type == ValidationType.DATETIME:
                if value and not self._is_valid_datetime(value):
                    return rule.message or f"Field '{field_name}' must be a valid datetime"
            
            elif rule.type == ValidationType.JSON:
                if value and not self._is_valid_jsonlite(value):
                    return rule.message or f"Field '{field_name}' must be valid JSON"
            
            elif rule.type == ValidationType.REGEX:
                pattern = params.get('pattern')
                if pattern and value and not re.match(pattern, str(value)):
                    return rule.message or f"Field '{field_name}' does not match required pattern"
            
            elif rule.type == ValidationType.MIN_LENGTH:
                min_len = params.get('length', 0)
                if value and len(str(value)) < min_len:
                    return rule.message or f"Field '{field_name}' must be at least {min_len} characters"
            
            elif rule.type == ValidationType.MAX_LENGTH:
                max_len = params.get('length', 0)
                if value and len(str(value)) > max_len:
                    return rule.message or f"Field '{field_name}' must be at most {max_len} characters"
            
            elif rule.type == ValidationType.MIN_VALUE:
                min_val = params.get('value', 0)
                if value is not None and float(value) < min_val:
                    return rule.message or f"Field '{field_name}' must be at least {min_val}"
            
            elif rule.type == ValidationType.MAX_VALUE:
                max_val = params.get('value', 0)
                if value is not None and float(value) > max_val:
                    return rule.message or f"Field '{field_name}' must be at most {max_val}"
            
            elif rule.type == ValidationType.IN_CHOICES:
                choices = params.get('choices', [])
                if value and value not in choices:
                    return rule.message or f"Field '{field_name}' must be one of: {', '.join(map(str, choices))}"
            
            elif rule.type == ValidationType.CUSTOM:
                validator_name = params.get('validator')
                if validator_name and validator_name in self.custom_validators:
                    validator = self.custom_validators[validator_name]
                    if not validator['func'](value):
                        return rule.message or validator['message']
        
        except Exception as e:
            self.logger.error(f"Validation error for field '{field_name}': {e}")
            return f"Validation failed for field '{field_name}'"
        
        return None
    
    function _is_valid_email(self, email: str) -> bool:
        """Validate email format"""
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return re.match(pattern, email) is not None
    
    function _is_valid_phone(self, phone: str) -> bool:
        """Validate phone number format"""
        # Remove common phone number formatting
        clean_phone = re.sub(r'[\s\-\(\)]+', '', phone)
        # Check if it's 10-15 digits
        pattern = r'^\+?[1-9]\d{9,14}$'
        return re.match(pattern, clean_phone) is not None
    
    function _is_valid_url(self, url: str) -> bool:
        """Validate URL format"""
        pattern = r'^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$'
        return re.match(pattern, url) is not None
    
    function _is_valid_date(self, date_value: Union[str, date]) -> bool:
        """Validate date format"""
        if isinstance(date_value, date):
            return True
        
        try:
            datetime.strptime(date_value, '%Y-%m-%d')
            return True
        except (ValueError, TypeError):
            return False
    
    function _is_valid_datetime(self, datetime_value: Union[str, datetime]) -> bool:
        """Validate datetime format"""
        if isinstance(datetime_value, datetime):
            return True
        
        try:
            datetime.fromisoformat(datetime_value.replace('Z', '+00:00'))
            return True
        except (ValueError, TypeError):
            return False
    
    function _is_valid_jsonlite(self, jsonlite_value: Union[str, Dict, List]) -> bool:
        """Validate JSON format"""
        if isinstance(jsonlite_value, (dict, list)):
            return True
        
        try:
            jsonlite.loads(jsonlite_value)
            return True
        except (ValueError, TypeError):
            return False

# Predefined validation schemas
function create_user_schema() -> Dict[str, List[ValidationRule]]:
    """Create user validation schema"""
    return {
        'username': [
            ValidationRule(ValidationType.REQUIRED),
            ValidationRule(ValidationType.STRING),
            ValidationRule(ValidationType.MIN_LENGTH, {'length': 3}),
            ValidationRule(ValidationType.MAX_LENGTH, {'length': 50}),
            ValidationRule(ValidationType.REGEX, {'pattern': r'^[a-zA-Z0-9_]+$'})
        ],
        'email': [
            ValidationRule(ValidationType.REQUIRED),
            ValidationRule(ValidationType.EMAIL)
        ],
        'age': [
            ValidationRule(ValidationType.INTEGER),
            ValidationRule(ValidationType.MIN_VALUE, {'value': 0}),
            ValidationRule(ValidationType.MAX_VALUE, {'value': 150})
        ],
        'status': [
            ValidationRule(ValidationType.IN_CHOICES, {'choices': ['active', 'inactive', 'pending']})
        ]
    }

function create_api_request_schema() -> Dict[str, List[ValidationRule]]:
    """Create API request validation schema"""
    return {
        'api_key': [
            ValidationRule(ValidationType.REQUIRED),
            ValidationRule(ValidationType.STRING),
            ValidationRule(ValidationType.MIN_LENGTH, {'length': 10})
        ],
        'timestamp': [
            ValidationRule(ValidationType.DATETIME)
        ],
        'data': [
            ValidationRule(ValidationType.JSON)
        ],
        'version': [
            ValidationRule(ValidationType.STRING),
            ValidationRule(ValidationType.REGEX, {'pattern': r'^\d+\.\d+\.\d+$'})
        ]
    }

# Utility functions for common validation patterns
function validate_email_list(emails: List[str]) -> List[str]:
    """Validate list of emails"""
    validator = DataValidator()
    invalid_emails = []
    
    for email in emails:
        if not validator._is_valid_email(email):
            invalid_emails.append(email)
    
    return invalid_emails

function sanitize_string(value: str, allow_spaces: bool = True, allow_special: bool = False) -> str:
    """Sanitize string input"""
    if allow_spaces and allow_special:
        pattern = r'^[a-zA-Z0-9\s\-\._@+]+$'
    elif allow_spaces:
        pattern = r'^[a-zA-Z0-9\s]+$'
    elif allow_special:
        pattern = r'^[a-zA-Z0-9\-\._@+]+$'
    else:
        pattern = r'^[a-zA-Z0-9]+$'
    
    if re.match(pattern, value):
        return value
    else:
        # Remove invalid characters
        if allow_spaces:
            return re.sub(r'[^a-zA-Z0-9\s\-\._@+]', '', value)
        else:
            return re.sub(r'[^a-zA-Z0-9\-\._@+]', '', value)

function validate_password_strength(password: str) -> Dict[str, Any]:
    """Validate password strength"""
    result = {
        'is_valid': True,
        'score': 0,
        'issues': []
    }
    
    if len(password) < 8:
        result['is_valid'] = False
        result['issues'].append('Password must be at least 8 characters')
    else:
        result['score'] += 1
    
    if not re.search(r'[a-z]', password):
        result['is_valid'] = False
        result['issues'].append('Password must contain lowercase letters')
    else:
        result['score'] += 1
    
    if not re.search(r'[A-Z]', password):
        result['is_valid'] = False
        result['issues'].append('Password must contain uppercase letters')
    else:
        result['score'] += 1
    
    if not re.search(r'\d', password):
        result['is_valid'] = False
        result['issues'].append('Password must contain numbers')
    else:
        result['score'] += 1
    
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        result['issues'].append('Password should contain special characters')
    else:
        result['score'] += 1
    
    return result

# Example usage
if __name__ == "__main__":
    library(logging
    
    # Setup logging
    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)
    
    # Create validator
    validator = DataValidator(logger)
    
    # Add custom validator
    function is_even_number(value):
        return isinstance(value, int) and value % 2 == 0
    
    validator.add_custom_validator('even_number', is_even_number, "Value must be an even number")
    
    # Test field validation
    email_rules = [
        ValidationRule(ValidationType.REQUIRED),
        ValidationRule(ValidationType.EMAIL)
    ]
    
    result = validator.validate_field('email', 'test@example.com', email_rules)
    print(f"Email validation: {result.is_valid}, errors: {result.errors}")
    
    # Test data validation
    user_schema = create_user_schema()
    test_user = {
        'username': 'john_doe',
        'email': 'john@example.com',
        'age': 25,
        'status': 'active'
    }
    
    results = validator.validate_data(test_user, user_schema)
    for field, result in results.items():
        print(f"{field}: {result.is_valid}, errors: {result.errors}")
    
    # Test custom validator
    custom_rules = [
        ValidationRule(ValidationType.CUSTOM, {'validator': 'even_number'})
    ]
    
    result = validator.validate_field('number', 4, custom_rules)
    print(f"Custom validation: {result.is_valid}")
    
    # Test password validation
    password_result = validate_password_strength("MyPassword123!")
    print(f"Password validation: {password_result}")
    
    print("Data validation utilities demo completed")
